### PR TITLE
feat(on_demand): Support epm and eps operators

### LINKED ISSUE #56572: Support epm() and eps() for on-demand metrics

### PR DESCRIPTION
**Pull request by armenzg**
Fixes #56572

### CODE REVIEW

File: src/sentry/snuba/metrics/extraction.py [hunk 1]
```diff
@@ -620,13 +609,41 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
...
-# This is used to map a metric to a function which generates a specification
-_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {
-    "on_demand_failure_count": failure_tag_spec,
-    "on_demand_failure_rate": failure_tag_spec,
-    "on_demand_apdex": apdex_tag_spec,
+def epm_tag_spec(_1: Project, _2: Optional[str]) -> List[TagSpec]:
+    pass
+
+
+DERIVED_METRICS = {
```
Comment (armenzg) [inline src/sentry/snuba/metrics/extraction.py:616]: I used this to populate other dictionaries. Less than ideal but it reduces the places to add a new metric.
Let me know if you prefer a separate PR for the refactoring aspects.

File: src/sentry/snuba/metrics/extraction.py [hunk 2]
```diff
@@ -620,13 +609,41 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
     ]
 
 
-# This is used to map a metric to a function which generates a specification
-_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {
-    "on_demand_failure_count": failure_tag_spec,
-    "on_demand_failure_rate": failure_tag_spec,
-    "on_demand_apdex": apdex_tag_spec,
+def epm_tag_spec(_1: Project, _2: Optional[str]) -> List[TagSpec]:
+    pass
```
Comment (armenzg) [inline src/sentry/snuba/metrics/extraction.py:613]: I need to work on this. Any pointers are greatly appreciated.

File: src/sentry/snuba/metrics/fields/snql.py [hunk 3]
```diff
@@ -802,6 +802,20 @@ def max_timestamp(aggregate_filter, org_id, use_case_id, alias=None):
     return timestamp_column_snql("maxIf", aggregate_filter, org_id, use_case_id, alias)
 
 
+def on_demand_epm_snql_factory(
+    aggregate_filter: Function, org_id: int, use_case_id: UseCaseID, alias: str
+) -> Function:
+    """Divide the total count of events by the time period"""
+    # XXX: This whole thing is wrong. Just pushing the code so
+    return Function(
```
Comment (armenzg) [inline src/sentry/snuba/metrics/fields/snql.py:810]: I need to work on this.

File: src/sentry/snuba/metrics/extraction.py [hunk 4]
```diff
@@ -123,6 +123,8 @@
     "failure_count": "on_demand_failure_count",
     "failure_rate": "on_demand_failure_rate",
     "apdex": "on_demand_apdex",
+    "epm": "on_demand_epm",
+    "eps": "on_demand_eps",
```
Comment (jan-auer) [inline src/sentry/snuba/metrics/extraction.py:127]: Should you also add mappings for `tpm` and `tps` here, or is that converted somewhere upstream of the extraction module?

**Review by jan-auer — APPROVED** [review]
Elegant solution! 

The implementation looks good to me and test coverage seems sufficient. Considering that these functions cannot be used from alerts, this should be safe to merge.

File: src/sentry/snuba/metrics/extraction.py (hunk 4 above)
Comment (armenzg) [reply src/sentry/snuba/metrics/extraction.py:127]: `tpm` and `tps` cannot be used in widgets.
It's a concept within Performance transaction summaries.

**Pull request merged**